stdlib/Distributed/src/remotecall.jl

```diff
@@ -351,7 +361,10 @@ end
 
 channel_type(rr::RemoteChannel{T}) where {T} = T
 
-serialize(s::ClusterSerializer, f::Future) = serialize(s, f, f.v === nothing)
+function serialize(s::ClusterSerializer, f::Future)
+    v_cache = @atomic f.v
+    serialize(s, f, v_cache === nothing)
+end
 serialize(s::ClusterSerializer, rr::RemoteChannel) = serialize(s, rr, true)
 function serialize(s::ClusterSerializer, rr::AbstractRemoteRef, addclient)
     if addclient
```

Comment:
We need to change the Future serializer to exclude the new fields:
```suggestion
function serialize(s::ClusterSerializer, f::Future)
    serialize_type(s, typeof(f))
    serialize(s, f.where)
    serialize(s, remoteref_id(f))
    value = @atomic f.v
    if value === nothing
        p = worker_id_from_socket(s.io)
        (p !== rr.where) && send_add_client(rr, p)
    end
    serialize(s, value)
end
function serialize(s::ClusterSerializer, f::RemoteChannel)
    p = worker_id_from_socket(s.io)
    (p !== rr.where) && send_add_client(rr, p)
    invoke(serialize, Tuple{AbstractSerializer, Any}, s, f)
end
serialize(s::AbstractSerializer, ::AbstractLock) = error("Locks cannot be serialized")
```
and corresponding `deserialize` methods
```julia
function deserialize(s::ClusterSerializer, T::Type{<:Future})
    where = deserialize(s, f.where)::Int
    rid = deserialize(s, remoteref_id(f))::RRID
    value = deserialize(s, f.value)
    # 1) send_add_client() is not executed when the ref is being serialized
    #    to where it exists, hence do it here.
    # 2) If we have received a 'fetch'ed Future or if the Future ctor found an
    #    already 'fetch'ed instance in client_refs (Issue #25847), we should not
    #    track it in the backing RemoteValue store.
    if where == myid() && value === nothing
        add_client(rid, where)
    end
    return T(where, rid, value) # ctor adds to client_refs table
end
function deserialize(s::ClusterSerializer, T::Type{<:RemoteChannel})
    rr = invoke(deserialize, Tuple{ClusterSerializer, DataType}, s, t)::T # deserialize a placeholder object
    rid = remoteref_id(rr)
    if rr.where == myid()
        # send_add_client() is not executed when the ref is being
        # serialized to where it exists
        add_client(rid, myid())
    end
    # call ctor to make sure this rr gets added to the client_refs table
    return T(rr.where, rid)
end
```

Reply:
I've been trying to make it work, but without any success. 
If I serialize the fields manually as proposed by you at some point the code expects a `Future` structure and I'm not really sure how to handle that.

Anyway i was thinking about the issue and as of right now:
1. Serialization takes the full `Future` with the new lock and serializes it
2. Deserialization takes the serialized data and constructs a new `Future` based on it, but I'm unsure whether it takes the serialized lock data or creates a new one (constructors only ever create a new one, apart from the default)

I made the serialization now serialize a copy of the Future with a new unused lock, but I'm not sure if it's necessary. It used to work fine before any serialization change anyway

Reply:
@vtjnash I rebased and it's finally passing CI.
What do you think of the above changes?
I couldn't get the serialization of only the "safe" fields working, so I just added a full copy in the serializer.
I'm not sure it's necessary though as serializing a locked lock doesn't matter in the end as during deserialization a new lock is being constructed.
Or is there some reason not to serialize a locked lock?

Reply:
I am not sure why you were trying to use `invoke`, since there is definitely no hope of that approach working at all, while I had written the code just above here that I think should be roughly correct. Anyways, this approach now seems suitable enough so I have merged it.